iris_ugrid/regrid.py

```diff
@@ -0,0 +1,420 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Provides ESMF representations of grids/UGRID meshes and a modified regridder.
+"""
+import ESMF
+import numpy as np
+from numpy import ma
+import scipy.sparse
+import cartopy.crs as ccrs
+
+
+class MeshInfo:
+    """
+    This class holds information about Meshes in a form similar to UGRID.
+    It contains methods for translating this information into ESMF objects.
+    """
+
+    def __init__(
+        self,
+        node_coords,
+        face_node_connectivity,
+        node_start_index,
+        elem_start_index=0,
+        areas=None,
+    ):
+        """
+        Creates a MeshInfo object describing the geometry/topology of a UGRID
+        mesh.
+
+        Args:
+
+        * node_coords
+            An Nx2 numpy array describing the location of the nodes of the mesh.
+            node_coords[:,0] describes the longitudes in degrees and
+            node_coords[:,1] describes the latitides in degrees
+        * face_node_connectivity
+            A numpy masked array describing the face node connectivity of the
+            mesh. The unmasked points of face_node_connectivity[i] describe
+            which nodes are connected to the i'th face.
+        * node_start_index
+            An integer describing which index is considered the initial index
+            by face_node_connectivity. UGRID supports both 0 based and 1 based
+            indexing, so both must be accounted for here
+
+        Kwargs:
+
+        * elem_start_index
+            An integer describing what index should be considered by ESMF to be
+            the start index for describing its elements. This makes no
+            difference to the regridding calculation and will only affect the
+            intermediate ESMF objects, should the user need access to them.
+        * areas
+            either None or a numpy array describing the areas associated with
+            each face. If None, then ESMF will use its own calculated areas.
```

Comment:
Why doesn't this implement a `crs` keyword, like `GridInfo` ?
I get that having them related by inheritance may be undesirable, but I guess we are probably aiming that they should be "as similar as makes sense given the difference" ?

But, I do see that in this case, you have stated the "longitudes/latitudes in degrees" explicit :+1: